lib/pbench/server/api/resources/query_apis/iteration_samples.py

```diff
@@ -0,0 +1,240 @@
+from http import HTTPStatus
+from logging import Logger
+
+import requests
+from flask import jsonify
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.auth import Auth
+from flask_restful import Resource, abort
+from pbench.server.api.resources.query_apis.elasticsearch_api import Elasticsearch
+from pbench.server.api.resources.query_apis import (
+    CONTEXT,
+    ElasticBase,
+    JSON,
+    Parameter,
+    ParamType,
+    PostprocessError,
+    Schema,
+    UnauthorizedAccess,
+)
+from pbench.server.database.models.datasets import Dataset, Metadata
+
+
+class IterationSamples(ElasticBase):
+    """
+    Create iteration samples aggregation based on a given run id.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("start", ParamType.DATE, required=True),
+                Parameter("end", ParamType.DATE, required=True),
+                Parameter("run_id", ParamType.STRING, required=True),
+            ),
+        )
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this run id, and determine whether the
+        authenticated user has READ access to this dataset. (Currently, this
+        means the authenticated user is the owner of the dataset, or has ADMIN
+        role.)
+
+        If the user has authorization to read the dataset, return the Dataset
+        object as CONTEXT so that the postprocess operation can proceed to get all
+        the iteration aggregations.
+        """
+        run_id = client_json.get("run_id")
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.error("Run ID {} not found", run_id)
+            abort(HTTPStatus.NOT_FOUND, message="Dataset not found")
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                "Dataset owner ID {} cannot be found in Users", dataset.owner_id
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For Iteration samples, we check authorization against the ownership of the
+        # dataset that was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, dataset["access"])
```

Comment:
`dataset` is also a SQLAlchemy model object with attributes, so `dataset.access` would be better.